liblangutil: refactor SourceReferenceFormatter (split out extraction / line-cutting part).

christianparpart opened the pull request:
adds SourceReferenceExtractor API to liblangutil.

This is a prerequisite of #5511 - and can be merged early.

FYI: that API could receive get some tests. maybe in the PR or another?.

Commit:
liblangutil: extends CharStream to know about the respective (file-)name (and adapt codebase to it)

Commit:
liblangutil: adds ParserBase::charStream() accessor

Commit:
liblangutil: SourceLocation: adds (shared) pointer to underlying CharStream source, eliminating sourceName

Also, adapted affecting code to those changes.

Commit:
liblangutil: SourceLocation: do not store sourceName but pointer to underlying CharStream (which knows about the source's filename)

chriseth commented:
Why did you not modify SourceReferenceFormatter? I thought it was a refactor. Also there seem to be still some TODOs in the code.

christianparpart commented:
@chriseth, it is, but that's a splitout of the bigger PR. the code changes of the SourceReferenceFormatter are in the main PR. you asked me to split a few commits out.

Commit:
liblangutil: adds Scanner.charStream() accessor

Commit:
liblangutil: Scanner: remove superfluous sourceName field (it's in CharStream already)

Also, ParserBase::sourceName() was dead code. Eliminating it should
increase test coverage (how sneaky) :-)

christianparpart commented:
due to rebase-fuckups with #5537, I cannot push onto the remote branch anymore. Have to start over, sorry.
